GPSSetup/clients/GPSSetup/controllers/client/customCode.js

```diff
@@ -0,0 +1,357 @@
+
+
+        // draw all shapes
+        var l = shapes.length;
+
+        // if shape length is 4 then just draw gps ZONE
+        // if shape length is 8 then also draw the traking zone
+
+        console.log("shape length" + l);
+
+        // draw the first shape
+        if (l >= 4){
+            ctx.beginPath();
+            for (var i = 0; i < 4; i++) {
+                var shape = shapes[i];
+                // We can skip the drawing of elements that have moved off the screen:
+                if (shape.x > this.width || shape.y > this.height ||
+                    shape.x + shape.w < 0 || shape.y + shape.h < 0) continue;
+
+                shapes[i].draw(ctx);
+
+                if(i === 0)
+                    ctx.moveTo(shape.x, shape.y);
+                else if (i < 4)
+                    ctx.lineTo(shape.x, shape.y);
+            }
+            ctx.closePath();
+            // ctx.fillStyle = '#f00';
+            ctx.fillStyle = "rgba(193, 66, 66, 0.4)";
+            //ctx.globalAlpha = 0.5;
+            ctx.fill();
+        }
+
+        // draw the second shape
+        if (l >= 8){
+            ctx.beginPath();
+            for (var i = 4; i < 8; i++) {
+                var shape = shapes[i];
+                // We can skip the drawing of elements that have moved off the screen:
+                if (shape.x > this.width || shape.y > this.height ||
+                    shape.x + shape.w < 0 || shape.y + shape.h < 0) continue;
+
+                shapes[i].draw(ctx);
+
+                if(i === 4)
+                    ctx.moveTo(shape.x, shape.y);
+                else if (i < 8)
+                    ctx.lineTo(shape.x, shape.y);
+            }
+            ctx.closePath();
+            // ctx.fillStyle = '#f00';
+            ctx.fillStyle = "rgba(66, 66, 193, 0.7)";
+            //ctx.globalAlpha = 0.5;
+            ctx.fill();        
+        }
+
+        // draw selection
+        // right now this is just a stroke along the edge of the selected Shape
+        if (this.selection !== null) {
+            ctx.strokeStyle = this.selectionColor;
+            ctx.lineWidth = this.selectionWidth;
+            var mySel = this.selection;
+            ctx.strokeRect(mySel.x,mySel.y,mySel.w,mySel.h);
+        }
+
+        // ** Add stuff you want drawn on top all the time here **
+
+        this.valid = true;
+    }
+};
+
+
+// Creates an object with x and y defined, set to the mouse position relative to the state's canvas
+// If you wanna be super-correct this can be tricky, we have to worry about padding and borders
+CanvasState.prototype.getMouse = function(e) {
+    console.log("CALLED: CanvasState.prototype.getMouse");
+    var element = this.canvas, offsetX = 0, offsetY = 0, mx, my;
+
+    // Compute the total offset
+    if (element.offsetParent !== undefined) {
+        do {
+            offsetX += element.offsetLeft;
+            offsetY += element.offsetTop;
+        } while ((element = element.offsetParent));
```

Comment:
Any reason why this can't simply be a while loop? Doesn't look like the "do" block affects the outcome of the while condition.